Mathlib/RingTheory/Polynomial/DegreeLT.lean

```diff
@@ -0,0 +1,238 @@
+
+-/
+
+
+namespace Polynomial
+
+/-- R[x]_n is notation for the submodule of polynomials of degree strictly less than n. -/
+scoped notation:9000 R "[x]_" n => Polynomial.degreeLT R n
+
+section degreeLT
+
+variable {R : Type*} [Semiring R] {m n : ℕ} (i : Fin n) (P : R[x]_n)
+
+variable (R) in
+/-- Basis for R[x]_n given by the `X^i` with `i < n`. -/
+noncomputable def degreeLT.basis (n : ℕ) : Basis (Fin n) R R[x]_n :=
+  Basis.ofEquivFun (Polynomial.degreeLTEquiv R n)
+
+instance : Module.Free R R[x]_n :=
+  Module.Free.of_basis <| degreeLT.basis R n
+
+lemma degreeLT_X_pow_mem (i : Fin n) : X ^ i.val ∈ R[x]_n := by
+  nontriviality R using mem_degreeLT, WithBot.bot_lt_iff_ne_bot
+  simp only [Polynomial.mem_degreeLT, Polynomial.degree_X_pow, Nat.cast_lt, Fin.is_lt]
+
+@[simp] lemma degreeLT.basis_val (i : Fin n) :
+    (degreeLT.basis R n i : R[X]) = X ^ i.val := by
+  simp only [degreeLT.basis, degreeLTEquiv, Basis.coe_ofEquivFun,
+    LinearEquiv.coe_symm_mk]
+  rw [Finset.sum_eq_single i (by aesop) (by aesop),
+      Pi.single_eq_same, monomial_one_right_eq_X_pow]
+
+lemma degreeLT.basis_apply (i : Fin n) :
+    degreeLT.basis R n i = ⟨X ^ i.val, degreeLT_X_pow_mem i⟩ :=
+  Subtype.ext <| degreeLT.basis_val i
+
+@[simp] lemma degreeLT.basis_repr : (degreeLT.basis R n).repr P i = (P : R[X]).coeff i :=
+  Basis.ofEquivFun_repr_apply _ _ _
+
+instance : Module.Finite R R[x]_n :=
+  Module.Finite.of_basis (degreeLT.basis _ _)
+
+variable (R) in
+/-- Basis for `(R[x]_m) × (R[x]_n)`. -/
+noncomputable def degreeLT.basis_prod (m n : ℕ) : Basis (Fin (m + n)) R ((R[x]_m) × (R[x]_n)) :=
+  ((degreeLT.basis R m).prod (degreeLT.basis R n)).reindex finSumFinEquiv
+
+@[simp] lemma degreeLT.basis_prod_castAdd (m n : ℕ) (i : Fin m) :
+    basis_prod R m n (i.castAdd n) = (⟨X ^ i.val, degreeLT_X_pow_mem i⟩, 0) := by
+  rw [basis_prod, Basis.reindex_apply, finSumFinEquiv_symm_apply_castAdd]
+  ext
+  · rw [Basis.prod_apply_inl_fst, basis_apply]
+  · rw [Basis.prod_apply_inl_snd]
+
+@[simp] lemma degreeLT.basis_prod_natAdd (m n : ℕ) (i : Fin n) :
+    basis_prod R m n (i.natAdd m) = (0, ⟨X ^ i.val, degreeLT_X_pow_mem i⟩) := by
+  rw [basis_prod, Basis.reindex_apply, finSumFinEquiv_symm_apply_natAdd]
+  ext
+  · rw [Basis.prod_apply_inr_fst]
+  · rw [Basis.prod_apply_inr_snd, basis_apply]
+
+/-- An isomorphism between `(R[x]_(n+m))` and `(R[x]_n) × (R[x]_m)` given by the fact that
+the bases are both indexed by `Fin (n+m)`. -/
+noncomputable def degreeLT.addLinearEquiv {n m : ℕ} :
+    (R[x]_(n + m)) ≃ₗ[R] (R[x]_n) × (R[x]_m) :=
+  Basis.equiv (degreeLT.basis _ _) (degreeLT.basis_prod _ _ _) (Equiv.refl _)
+
+@[aesop unsafe 50%]
+theorem degree_add_lt_of_degree_lt {p q : R[X]} {n : ℕ}
+    (hp : degree p < n) (hq : degree q < n) :
+    degree (p + q) < n :=
+  (degree_add_le p q).trans_lt <| max_lt hp hq
+
+@[aesop unsafe 50%]
+theorem degree_mul_lt_of_lt_left {p q : R[X]} {a : WithBot ℕ} {b : ℕ}
```

Comment:
I am confused about why you allow a to be bot but not b, when you have degree < a but only <= b.